src/udev/udev-rules.c

```
@@ -1574,6 +1583,19 @@ static int udev_rule_apply_token_to_event(
                         val = hashmap_get(properties_list, token->data);
 
                 return token_match_string(token, val);
+        case TK_M_CONST: {
+                const char *k = token->data;
+
+                if (streq(k, "arch"))
+                        val = architecture_to_string(uname_architecture());
+                else if (streq(k, "virt"))
+                        val = virtualization_to_string(detect_virtualization());
+                else {
+                        log_rule_warning(dev, rules, "Unknown key '%s' will never match", k);
+                        return OP_NOMATCH;
+                }
```

I think the check should be done in `parse_token()`, instead of here.

---

Right, that's better. Fail early, etc.

So just `assert` here.

---

Yes, that seems to be a good idea. @keszybz What do you think?